blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineStepVisitor.java

```diff
@@ -209,8 +210,9 @@ public void atomNode(@CheckForNull FlowNode before, @Nonnull FlowNode atomNode,
             }
             stepMap.put(node.getId(), node);
 
-            //If there is closest block boundary, we capture it's error to the last step encountered and prepare for next block.
-            if(closestEndNode!=null && closestEndNode.getError() != null) {
+            // If there is closest block boundary, we capture it's error to the last step encountered and prepare for next block.
+            // If the previous step is in error, don't mark this step with a failure
+            if(closestEndNode!=null && closestEndNode.getError() != null && new NodeRunStatus(before).result != BlueRunResult.FAILURE) {
```

Comment:
If the previous step failed for a non-syntax reason, it would mark the current step as a failure.

For example:
```
try {
  sh 'foo bar'
} catch (e) {
   echo 'would be marked as failed'
}
```